### ISSUE TITLE
grid_sample() and align_corners parameter - new default value in pytorch v1.4.0

### ISSUE THREAD

**Issue by poxyu** [kornia/kornia#388]
It is said pytorch devs are going to change the default behavior of `grid_sample()` in pytorch v1.4.0:
from `align_corners=True` to `align_corners=False`

Here is the link:
https://pytorch.org/docs/stable/nn.functional.html#grid-sample

<img width="679" alt="Снимок экрана 2019-12-22 в 2 39 28" src="https://user-images.githubusercontent.com/8268361/71315867-07d74000-2467-11ea-9441-1e016dbfac8a.png">

What do you think about explicitly adding `align_corners=True` everywhere in `grid_sample()` so as not to ruin the current behavior of kornia (and your tests of course)?

**Comment by edgarriba** [comment]
yes, we could do that. We should also check the performance when the flag is set to False.

**Comment by edgarriba** [comment]
pytorch 1.4 is here. I will update at some point during this weekend. What should we do here ? /cc @ducha-aiki

**Comment by poxyu** [comment]
I've read several times docs to `torch.nn.functional.interpolate` and `torch.nn.functional.grid_sample`.
After this I decided to... use `align_corners = False` :joy:

Main reason is `making the sampling more resolution agnostic`. And it makes more sense to me.

Here is my ugly solution for now:
```
def create_meshgrid_not_aligned_corners(
        height: int,
        width: int,
        device: torch.device = torch.device("cuda")
) -> torch.Tensor:
    """Generates a coordinate grid for an image.
    The grid is normalized to be in the range [-1,1] to be consistent with the pytorch
    function grid_sample with align_corners = False (IMPORTANT)
    http://pytorch.org/docs/master/nn.html#torch.nn.functional.grid_sample
    Args:
        height (int): the image height (rows).
        width (int): the image width (cols).
    Return:
        torch.Tensor: returns a grid tensor with shape :math:`(1, H, W, 2)`.
    """

    start = -1.
    stop = 1.

    x_half_pixel_size = (stop - start) / width / 2.
    y_half_pixel_size = (stop - start) / height / 2.

    x_start = start + x_half_pixel_size
    x_stop = stop - x_half_pixel_size

    y_start = start + y_half_pixel_size
    y_stop = stop - y_half_pixel_size
        
    # generate coordinates
    xs = torch.linspace(x_start, x_stop, width, device=device, dtype=torch.float32)
    ys = torch.linspace(y_start, y_stop, height, device=device, dtype=torch.float32)
    # generate grid by stacking coordinates
    base_grid: torch.Tensor = torch.stack(
        torch.meshgrid([xs, ys])).transpose(1, 2)  # 2xHxW
    return torch.unsqueeze(base_grid, dim=0).permute(0, 2, 3, 1)  # 1xHxWx2
```

**Comment by edgarriba** [comment]
@poxyu makes sense. Actually we noticed some instabilities in the scale space algorithm and suspecting that this patch could help to improve. Could you try to make a PR with it ? By now, I hacked all the parameters in grid_sample to make it compliant with the old versions so that tests and behavior is maintained. See this: https://github.com/kornia/kornia/pull/402

**Comment by poxyu** [comment]
I can add additional parameter to `kornia.utils.grid.create_meshgrid` (and probably to `create_meshgrid3d`).
For example, `align_corners: Optional[bool] = True` (or False?) and make it work with `normalized_coordinates`, something like:
```
def create_meshgrid(
        height: int,
        width: int,
        normalized_coordinates: Optional[bool] = True,
        align_corners: Optional[bool] = True,
        device: Optional[torch.device] = torch.device('cpu')
) -> torch.Tensor:
	...

	if normalized_coordinates:
		if align_corners:
			# leave current logic
		else:
			# here goes my patch
	...
```
What do you think @edgarriba?

**Comment by poxyu** [comment]
When you have time, just let me know what you think

**Comment by edgarriba** [comment]
@poxyu yes please go ahead, just make sure tests are provided so that we can Keep backward compatibility with the current behaviour.

**Comment by edgarriba** [comment]
@poxyu maybe time for this refactoring ?

**Comment by poxyu** [comment]
@edgarriba  I just don't have time right now to write tests. But the main part is ready and used since that time... 
```
def create_meshgrid(
        height: int,
        width: int,
        normalized_coordinates: Optional[bool] = True,
        align_corners: Optional[bool] = True,
        device: Optional[torch.device] = torch.device("cpu")) -> torch.Tensor:
    """Generates a coordinate grid for an image.
    When the flag ``normalized_coordinates`` is set to True, the grid is
    normalized to be in the range [-1, 1] to be consistent with the pytorch
    function grid_sample.
    http://pytorch.org/docs/master/nn.html#torch.nn.functional.grid_sample
    Args:
        height (int): the image height (rows).
        width (int): the image width (cols).
        normalized_coordinates (Optional[bool]): whether to normalize
          coordinates in the range [-1, 1] in order to be consistent with the
          PyTorch function grid_sample.
        align_corners (Optional[bool]): if set to ``True``, the extrema [-1, 1]
          are considered as referring to the center points of the input's corner pixels;
          if set to ``False``, they are instead considered as referring to the corner points
          of the input's corner pixels, making the sampling more resolution agnostic 
          (if ``normalized_coordinates`` is set to ``True``).
    Return:
        torch.Tensor: returns a grid tensor with shape :math:`(1, H, W, 2)`.
    """
    # generate coordinates
    xs: Optional[torch.Tensor] = None
    ys: Optional[torch.Tensor] = None
    if normalized_coordinates:
        # extrema [-1, 1]
        x_start, x_stop = -1., 1.
        y_start, y_stop = -1., 1.
        # calculate corner points
        if not align_corners:
            x_half_pixel_size = (x_stop - x_start) / float(width) / 2.
            y_half_pixel_size = (y_stop - y_start) / float(height) / 2.
            # update extrema
            x_start, x_stop = x_start + x_half_pixel_size, x_stop - x_half_pixel_size
            y_start, y_stop = y_start + y_half_pixel_size, y_stop - y_half_pixel_size
        # generate coordinates
        xs = torch.linspace(x_start, x_stop, width, device=device, dtype=torch.float)
        ys = torch.linspace(y_start, y_stop, height, device=device, dtype=torch.float)
    else:
        # generate coordinates
        xs = torch.linspace(0, width - 1, width, device=device, dtype=torch.float)
        ys = torch.linspace(0, height - 1, height, device=device, dtype=torch.float)
    # generate grid by stacking coordinates
    base_grid: torch.Tensor = torch.stack(
        torch.meshgrid([xs, ys])).transpose(1, 2)  # 2xHxW
    return torch.unsqueeze(base_grid, dim=0).permute(0, 2, 3, 1)  # 1xHxWx2
```

**Comment by edgarriba** [comment]
@poxyu awesome, maybe with just the tests we already have should be enough. @ducha-aiki what do you think ?
